File: tools/local/history.py
from __future__ import annotations

import argparse
import json
import os
import time
from datetime import date, datetime, timedelta
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen

ROOT = Path('.local-data')
OBJECTS = ROOT / 'objects'
PARQUET_ROOT = OBJECTS / 'prices' / 'daily'
JSON_ROOT = OBJECTS / 'prices' / 'daily-json'
CLOSED_ROOT = ROOT / 'history-closed'
LATEST_FEATURE = OBJECTS / 'metadata' / 'latest-feature-state.json'
MASSIVE_BASE_URL = 'https://api.massive.com'
DEFAULT_REQUIRED_SESSIONS = 337
PACE_SECONDS = 13.0
MAX_RETRIES = 4


def load_dev_vars() -> dict[str, str]:
    path = Path('.dev.vars')
    if not path.exists():
        return {}
    out: dict[str, str] = {}
    for raw in path.read_text().splitlines():
        line = raw.strip()
        if not line or line.startswith('#') or '=' not in line:
            continue
        key, value = line.split('=', 1)
        value = value.strip().strip('"').strip("'")
        out[key.strip()] = value
    return out


def massive_api_key() -> str | None:
    return os.getenv('MASSIVE_API_KEY') or load_dev_vars().get('MASSIVE_API_KEY')


def latest_completed_session() -> date:
    if not LATEST_FEATURE.exists():
        raise SystemExit('latest feature state missing; run npm run local:sync first or pass --end=YYYY-MM-DD')
    payload = json.loads(LATEST_FEATURE.read_text())
    value = payload.get('as_of') or payload.get('date')
    if not value:
        raise SystemExit('latest feature state has no as_of/date; pass --end=YYYY-MM-DD')
    return date.fromisoformat(str(value))


def json_path(day: date) -> Path:
    return JSON_ROOT / f'date={day.isoformat()}' / 'bars.json'


def parquet_path(day: date) -> Path:
    return PARQUET_ROOT / f'date={day.isoformat()}' / 'bars.parquet'


def closed_path(day: date) -> Path:
    return CLOSED_ROOT / f'{day.isoformat()}.json'


def normalize_document(day: date, rows: list[dict[str, object]], *, source: str, producer: str) -> dict[str, object]:
    bars: list[dict[str, object]] = []
    for row in rows:
        symbol = row.get('symbol') or row.get('T')
        values = [row.get('open', row.get('o')), row.get('high', row.get('h')), row.get('low', row.get('l')), row.get('close', row.get('c')), row.get('volume', row.get('v'))]
        if not symbol or any(value is None for value in values):
            continue
        try:
            nums = [float(value) for value in values]
        except (TypeError, ValueError):
            continue
        bars.append({
            'symbol': str(symbol),
            'date': day.isoformat(),
            'open': nums[0],
            'high': nums[1],
            'low': nums[2],
            'close': nums[3],
            'volume': nums[4],
        })
    return {
        'date': day.isoformat(),
        'adjusted': True,
        'source': source,
        'producer': producer,
        'count': len(bars),
        'bars': bars,
    }


def write_json_document(day: date, payload: dict[str, object]) -> None:
    path = json_path(day)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(payload, separators=(',', ':')))


def materialize_parquet(day: date) -> bool:
    source = parquet_path(day)
    target = json_path(day)
    if target.exists():
        return True
    if not source.exists():
        return False
    try:
        import polars as pl
    except ImportError as exc:
        raise SystemExit('polars is required to materialize existing parquet history; run npm run local:history:setup') from exc
    frame = pl.read_parquet(source)
    payload = normalize_document(day, frame.to_dicts(), source='r2-parquet-mirror', producer='codespaces-history-materializer')
    if not payload['count']:
        raise RuntimeError(f'parquet file contained no valid bars: {source}')
    write_json_document(day, payload)
    return True


def existing_session_dates(end: date | None = None) -> list[date]:
    dates: set[date] = set()
    if JSON_ROOT.exists():
        for path in JSON_ROOT.glob('date=*/bars.json'):
            try:
                day = date.fromisoformat(path.parent.name.removeprefix('date='))
            except ValueError:
                continue
            if end is not None and day > end:
                continue
            try:
                payload = json.loads(path.read_text())
            except Exception:
                continue
            if payload.get('bars'):
                dates.add(day)
    return sorted(dates)


def materialize_all_parquet(end: date | None = None) -> int:
    if not PARQUET_ROOT.exists():
        return 0
    converted = 0
    for path in sorted(PARQUET_ROOT.glob('date=*/bars.parquet')):
        try:
            day = date.fromisoformat(path.parent.name.removeprefix('date='))
        except ValueError:
            continue
        if end is not None and day > end:
            continue
        if json_path(day).exists():
            continue
        if materialize_parquet(day):
            converted += 1
            if converted % 25 == 0:
                print(f'[local:history] materialized {converted} parquet sessions')
    return converted


def massive_grouped_daily(day: date, api_key: str) -> dict[str, object]:
    params = urlencode({'adjusted': 'true', 'apiKey': api_key})
    url = f'{MASSIVE_BASE_URL}/v2/aggs/grouped/locale/us/market/stocks/{day.isoformat()}?{params}'
    request = Request(url, headers={'Accept': 'application/json', 'User-Agent': 'FutureView-History-Recovery/1.0'})
    for attempt in range(MAX_RETRIES):
        try:
            with urlopen(request, timeout=30) as response:  # noqa: S310
                payload = json.loads(response.read().decode('utf-8'))
            if not isinstance(payload, dict):
                raise RuntimeError('Massive returned non-object JSON')
            return payload
        except HTTPError as exc:
            body = exc.read().decode('utf-8', errors='replace')
            retryable = exc.code == 429 or 500 <= exc.code < 600
            if retryable and attempt + 1 < MAX_RETRIES:
                retry_after = exc.headers.get('Retry-After')
                delay = float(retry_after) if retry_after else 15.0 * (attempt + 1)
                print(f'[local:history] Massive HTTP {exc.code}; retry after {delay:.1f}s')
                time.sleep(delay)
                continue
            raise RuntimeError(f'Massive HTTP {exc.code} for {day}: {body[:300]}') from exc
        except URLError as exc:
            if attempt + 1 < MAX_RETRIES:
                delay = 2.0 * (attempt + 1)
                time.sleep(delay)
                continue
            raise RuntimeError(f'Massive transport error for {day}: {exc.reason}') from exc
    raise RuntimeError(f'Massive request exhausted retries for {day}')


def recover_until_sessions(required: int, end: date) -> tuple[int, int]:
    sessions = existing_session_dates(end)
    if len(sessions) >= required:
        return 0, len(sessions)
    api_key = massive_api_key()
    if not api_key:
        raise SystemExit(f'need {required - len(sessions)} more sessions but MASSIVE_API_KEY is unavailable')

    cursor = sessions[0] - timedelta(days=1) if sessions else end
    fetched = 0
    while len(sessions) < required:
        if cursor.weekday() >= 5:
            cursor -= timedelta(days=1)
            continue
        if json_path(cursor).exists():
            try:
                payload = json.loads(json_path(cursor).read_text())
                if payload.get('bars'):
                    sessions.append(cursor)
            except Exception:
                pass
            cursor -= timedelta(days=1)
            continue
        if materialize_parquet(cursor):
            sessions.append(cursor)
            cursor -= timedelta(days=1)
            continue
        if closed_path(cursor).exists():
            cursor -= timedelta(days=1)
            continue

        payload = massive_grouped_daily(cursor, api_key)
        raw_rows = payload.get('results') if isinstance(payload.get('results'), list) else []
        document = normalize_document(cursor, raw_rows, source='massive', producer='codespaces-history-recovery')
        fetched += 1
        if document['count']:
            write_json_document(cursor, document)
            sessions.append(cursor)
            sessions.sort()
            print(f"[local:history] recovered {cursor}: {document['count']} bars ({len(sessions)}/{required} sessions)")
        else:
            CLOSED_ROOT.mkdir(parents=True, exist_ok=True)
            closed_path(cursor).write_text(json.dumps({'date': cursor.isoformat(), 'status': 'no_market_data', 'checked_at': datetime.utcnow().isoformat() + 'Z'}))
            print(f'[local:history] no market data {cursor}; marked closed/unavailable')
        cursor -= timedelta(days=1)
        if len(sessions) < required:
            time.sleep(PACE_SECONDS)
    return fetched, len(sessions)


def main() -> None:
    parser = argparse.ArgumentParser(description='Build/recover canonical local daily history without coupling backtests to Massive')
    parser.add_argument('--mode', choices=('ensure', 'materialize'), default='ensure')
    parser.add_argument('--sessions', type=int, default=DEFAULT_REQUIRED_SESSIONS)
    parser.add_argument('--end', type=date.fromisoformat, default=None)
    args = parser.parse_args()
    if args.sessions < 1:
        raise SystemExit('--sessions must be positive')
    end = args.end or latest_completed_session()

    print(f'[local:history] target end={end} mode={args.mode}')
    converted = materialize_all_parquet(end)
    print(f'[local:history] parquet materialized: {converted}')
    if args.mode == 'materialize':
        sessions = existing_session_dates(end)
        print(f'[local:history] READY: {len(sessions)} local sessions through {end}')
        return

    fetched, total = recover_until_sessions(args.sessions, end)
    sessions = existing_session_dates(end)
    print('\n[local:history] READY')
    print(f'[local:history] sessions: {total} ({sessions[-args.sessions]} -> {sessions[-1]})')
    print(f'[local:history] Massive recovery requests this run: {fetched}')
    print(f'[local:history] canonical history: {JSON_ROOT}/')


if __name__ == '__main__':
    main()

File: tools/local/history.mjs
import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import path from "node:path";

import { asyncBufferFromFile, parquetReadObjects } from "hyparquet";

const ROOT = ".local-data";
const OBJECTS = path.join(ROOT, "objects");
const PARQUET_ROOT = path.join(OBJECTS, "prices", "daily");
const JSON_ROOT = path.join(OBJECTS, "prices", "daily-json");
const CLOSED_ROOT = path.join(ROOT, "history-closed");
const LATEST_FEATURE = path.join(OBJECTS, "metadata", "latest-feature-state.json");
const MASSIVE_BASE_URL = "https://api.massive.com";
const DEFAULT_REQUIRED_SESSIONS = 337;
const PACE_MS = 13_000;
const MAX_RETRIES = 4;

function fail(message) {
  console.error(`\n[local:history] ERROR: ${message}`);
  process.exit(1);
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function parseArgs() {
  const out = { mode: "ensure", sessions: DEFAULT_REQUIRED_SESSIONS, end: null };
  for (const arg of process.argv.slice(2)) {
    if (arg.startsWith("--mode=")) out.mode = arg.slice(7);
    else if (arg.startsWith("--sessions=")) out.sessions = Number(arg.slice(11));
    else if (arg.startsWith("--end=")) out.end = arg.slice(6);
    else fail(`unknown argument: ${arg}`);
  }
  if (!new Set(["ensure", "materialize"]).has(out.mode)) fail("--mode must be ensure or materialize");
  if (!Number.isInteger(out.sessions) || out.sessions < 1) fail("--sessions must be a positive integer");
  if (out.end && !/^\d{4}-\d{2}-\d{2}$/.test(out.end)) fail("--end must be YYYY-MM-DD");
  return out;
}

function loadDevVars() {
  if (!existsSync(".dev.vars")) return {};
  const out = {};
  for (const raw of readFileSync(".dev.vars", "utf8").split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#")) continue;
    const idx = line.indexOf("=");
    if (idx < 1) continue;
    let value = line.slice(idx + 1).trim();
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }
    out[line.slice(0, idx).trim()] = value;
  }
  return out;
}

function massiveApiKey() {
  return process.env.MASSIVE_API_KEY || loadDevVars().MASSIVE_API_KEY || null;
}

function shiftDate(iso, days) {
  const value = new Date(`${iso}T12:00:00Z`);
  value.setUTCDate(value.getUTCDate() + days);
  return value.toISOString().slice(0, 10);
}

function isWeekend(iso) {
  const day = new Date(`${iso}T12:00:00Z`).getUTCDay();
  return day === 0 || day === 6;
}

function latestCompletedSession() {
  if (!existsSync(LATEST_FEATURE)) fail("latest feature state missing; run npm run local:sync first or pass --end=YYYY-MM-DD");
  const payload = JSON.parse(readFileSync(LATEST_FEATURE, "utf8"));
  const value = payload.as_of ?? payload.date ?? null;
  if (!value || !/^\d{4}-\d{2}-\d{2}$/.test(String(value))) fail("latest feature state has no valid as_of/date; pass --end=YYYY-MM-DD");
  return String(value);
}

function jsonPath(date) {
  return path.join(JSON_ROOT, `date=${date}`, "bars.json");
}

function parquetPath(date) {
  return path.join(PARQUET_ROOT, `date=${date}`, "bars.parquet");
}

function closedPath(date) {
  return path.join(CLOSED_ROOT, `${date}.json`);
}

function normalizeDocument(date, rows, { source, producer }) {
  const bars = [];
  for (const row of rows) {
    const symbol = row?.symbol ?? row?.T ?? row?.ticker ?? null;
    const rawValues = [
      row?.open ?? row?.o,
      row?.high ?? row?.h,
      row?.low ?? row?.l,
      row?.close ?? row?.c,
      row?.volume ?? row?.v,
    ];
    if (!symbol || rawValues.some((value) => value == null || value === "")) continue;
    const values = rawValues.map(Number);
    if (!values.every(Number.isFinite)) continue;
    bars.push({
      symbol: String(symbol),
      date,
      open: values[0],
      high: values[1],
      low: values[2],
      close: values[3],
      volume: values[4],
    });
  }
  return {
    date,
    adjusted: true,
    source,
    producer,
    count: bars.length,
    bars,
  };
}

function writeJsonDocument(date, payload) {
  const file = jsonPath(date);
  mkdirSync(path.dirname(file), { recursive: true });
  writeFileSync(file, JSON.stringify(payload));
}

async function materializeParquet(date) {
  const source = parquetPath(date);
  const target = jsonPath(date);
  if (existsSync(target)) return true;
  if (!existsSync(source)) return false;

  const file = await asyncBufferFromFile(source);
  const rows = await parquetReadObjects({ file });
  const payload = normalizeDocument(date, rows, {
    source: "r2-parquet-mirror",
    producer: "codespaces-history-materializer",
  });
  if (!payload.count) throw new Error(`parquet file contained no valid bars: ${source}`);
  writeJsonDocument(date, payload);
  return true;
}

function dateDirectories(root, fileName) {
  if (!existsSync(root)) return [];
  return readdirSync(root, { withFileTypes: true })
    .filter((entry) => entry.isDirectory() && /^date=\d{4}-\d{2}-\d{2}$/.test(entry.name))
    .map((entry) => ({ date: entry.name.slice(5), file: path.join(root, entry.name, fileName) }))
    .filter((entry) => existsSync(entry.file))
    .sort((a, b) => a.date.localeCompare(b.date));
}

function existingSessionDates(end = null) {
  const dates = [];
  for (const entry of dateDirectories(JSON_ROOT, "bars.json")) {
    if (end && entry.date > end) continue;
    try {
      const payload = JSON.parse(readFileSync(entry.file, "utf8"));
      if (Array.isArray(payload.bars) && payload.bars.length) dates.push(entry.date);
    } catch {
      // Ignore unreadable partial/corrupt local cache; recovery can replace it.
    }
  }
  return dates;
}

async function materializeAllParquet(end = null) {
  let converted = 0;
  for (const entry of dateDirectories(PARQUET_ROOT, "bars.parquet")) {
    if (end && entry.date > end) continue;
    if (existsSync(jsonPath(entry.date))) continue;
    if (await materializeParquet(entry.date)) {
      converted += 1;
      if (converted % 25 === 0) console.log(`[local:history] materialized ${converted} parquet sessions`);
    }
  }
  return converted;
}

async function massiveGroupedDaily(date, apiKey) {
  const url = new URL(`/v2/aggs/grouped/locale/us/market/stocks/${date}`, MASSIVE_BASE_URL);
  url.searchParams.set("adjusted", "true");
  url.searchParams.set("apiKey", apiKey);

  for (let attempt = 0; attempt < MAX_RETRIES; attempt += 1) {
    let response;
    try {
      response = await fetch(url, {
        headers: { Accept: "application/json", "User-Agent": "FutureView-History-Recovery/2.0" },
        signal: AbortSignal.timeout(30_000),
      });
    } catch (error) {
      if (attempt + 1 >= MAX_RETRIES) throw new Error(`Massive transport error for ${date}: ${error.message}`);
      const delay = 2_000 * (attempt + 1);
      console.warn(`[local:history] Massive transport error; retry after ${(delay / 1000).toFixed(1)}s`);
      await sleep(delay);
      continue;
    }

    if (response.ok) {
      const payload = await response.json();
      if (!payload || typeof payload !== "object") throw new Error("Massive returned non-object JSON");
      return payload;
    }

    const body = await response.text();
    const retryable = response.status === 429 || response.status >= 500;
    if (!retryable || attempt + 1 >= MAX_RETRIES) {
      throw new Error(`Massive HTTP ${response.status} for ${date}: ${body.slice(0, 300)}`);
    }
    const retryAfter = Number(response.headers.get("retry-after"));
    const delay = Number.isFinite(retryAfter) && retryAfter >= 0
      ? Math.max(1_000, retryAfter * 1000)
      : 15_000 * (attempt + 1);
    console.warn(`[local:history] Massive HTTP ${response.status}; retry after ${(delay / 1000).toFixed(1)}s`);
    await sleep(delay);
  }

  throw new Error(`Massive request exhausted retries for ${date}`);
}

async function recoverUntilSessions(required, end) {
  let sessions = existingSessionDates(end);
  if (sessions.length >= required) return { requests: 0, sessions: sessions.length };

  const apiKey = massiveApiKey();
  if (!apiKey) fail(`need ${required - sessions.length} more sessions but MASSIVE_API_KEY is unavailable`);

  let cursor = sessions.length ? shiftDate(sessions[0], -1) : end;
  let requests = 0;
  while (sessions.length < required) {
    if (isWeekend(cursor)) {
      cursor = shiftDate(cursor, -1);
      continue;
    }

    if (existsSync(jsonPath(cursor))) {
      try {
        const payload = JSON.parse(readFileSync(jsonPath(cursor), "utf8"));
        if (Array.isArray(payload.bars) && payload.bars.length) sessions.push(cursor);
      } catch {
        // Fall through and recover this session.
      }
      cursor = shiftDate(cursor, -1);
      continue;
    }

    if (await materializeParquet(cursor)) {
      sessions.push(cursor);
      cursor = shiftDate(cursor, -1);
      continue;
    }

    if (existsSync(closedPath(cursor))) {
      cursor = shiftDate(cursor, -1);
      continue;
    }

    const payload = await massiveGroupedDaily(cursor, apiKey);
    requests += 1;
    const rows = Array.isArray(payload.results) ? payload.results : [];
    const document = normalizeDocument(cursor, rows, {
      source: "massive",
      producer: "codespaces-history-recovery",
    });

    if (document.count) {
      writeJsonDocument(cursor, document);
      sessions.push(cursor);
      sessions.sort();
      console.log(`[local:history] recovered ${cursor}: ${document.count} bars (${sessions.length}/${required} sessions)`);
    } else {
      mkdirSync(CLOSED_ROOT, { recursive: true });
      writeFileSync(closedPath(cursor), JSON.stringify({
        date: cursor,
        status: "no_market_data",
        checked_at: new Date().toISOString(),
      }));
      console.log(`[local:history] no market data ${cursor}; marked closed/unavailable`);
    }

    cursor = shiftDate(cursor, -1);
    if (sessions.length < required) await sleep(PACE_MS);
  }

  return { requests, sessions: sessions.length };
}

const options = parseArgs();
const end = options.end ?? latestCompletedSession();
console.log(`[local:history] target end=${end} mode=${options.mode}`);

const converted = await materializeAllParquet(end);
console.log(`[local:history] parquet materialized: ${converted}`);

if (options.mode === "materialize") {
  const sessions = existingSessionDates(end);
  console.log(`[local:history] READY: ${sessions.length} local sessions through ${end}`);
  process.exit(0);
}

const result = await recoverUntilSessions(options.sessions, end);
const sessions = existingSessionDates(end);
console.log("\n[local:history] READY");
console.log(`[local:history] sessions: ${result.sessions} (${sessions.at(-options.sessions)} -> ${sessions.at(-1)})`);
console.log(`[local:history] Massive recovery requests this run: ${result.requests}`);
console.log(`[local:history] canonical history: ${JSON_ROOT}/`);
